### ISSUE TITLE
Cannot parse non-interactive ripgrep v12.1.1 output

### ISSUE THREAD

**Issue by aaronbieber** [Kinneyzhang/gk-roam.el#2]
I cannot get gk-roam to work at all, which seems to be due to unexpected format of the `rg` output. The error that I get when running `gk-roam-update` is:

```
error in process sentinel: progn: Opening input file: Invalid argument, c:/Users/Aaron/Dropbox/org/gk-roam/20200906132737-Powerful-questions.org: [[file:20200906132600-Coaching.org][Coaching]]
```

It cannot open that file because the whole Org link text is included in the line. I started debugging and stepping through the processing of `rg` output and noticed a couple things that look wrong for my output, namely there are no line numbers in non-interactive `rg` output.

I'm not sure how your system is running the process, but this is what my output looks like:

```
c:\Users\Aaron\Dropbox\org\gk-roam\20200906132737-Powerful-questions.org: [[file:20200906132600-Coaching.org][Coaching]]
c:\Users\Aaron\Dropbox\org\gk-roam\20200906133044-Great-coaching-ideas.org: [[file:20200906132600-Coaching.org][Coaching]]
```

Note that if I run `rg` on an interactive command line, it looks different. Based on my reading of the code, the code may be expecting a format that looks like the interactive output, but that is not what elisp sees. To simulate non-interactive output, pipe through `less` or `more`.

I got these results with a download of the gk-roam source from GitHub as of yesterday, in Emacs 27.1, on Windows 10 Pro.

**Comment by Kinneyzhang** [comment]
I got no idea about this. In MacOS, the interactive output of rg in command line and the none-interactive one in Elisp are the same. I got rg output in Elisp by change the code to:

```
(generate-new-buffer-name "*gk-roam-rg*")
```
delete the blank before "*", you can see the output by switching to "*gk-roam-rg*" buffer.

Maybe, you could try to add "-n" to rg process parameter in Elisp code.

I don't use windows, cannot test for you. Sorry for that!

**Comment by aaronbieber** [comment]
I also have a Mac, I just don't use it as often. I got gk-roam set up there and confirmed exactly what you described.

I also tried in Ubuntu and had the same results as in Windows. Honestly, the result in macOS is confusing, because most command-line programs (including git and grep to name only two) have different output formats for interactive and non-interactive shells, and Emacs `start-process` should count as a non-interactive shell.

Unfortunately, the `-n` flag just inserts the line number in a different way, it doesn't result in the output that we're looking for. I'm going to dig a little deeper into why Emacs on macOS is producing interactive shell. The non-interactive shell output format is much easier to parse (which is why it exists; it is better for piping through other commands, etc.), so if there is a way to force that, I think that would be the best outcome.

**Comment by aaronbieber** [comment]
Apparently Emacs on macOS uses a terminal shell to run process commands, and Linux and Windows do not. This surprises me, because usually Windows is the operating system that does everything differently.

From the ripgrep man page:

```
       --no-heading
           Don't group matches by each file. If --no-heading is provided in addition to the -H/--with-filename
           flag, then file paths will be printed as a prefix for every matched line. This is the default mode
           when not printing to a terminal.

           This overrides the --heading flag.
```

Note *"this is the default mode when not printing to a terminal"*, which is what I would expect when spawning a process from within a programming language, like elisp. That said, we can easily pass this flag and get the same behavior on all platforms, and as a bonus, the output is easier to parse.

I'm going to work on a patch to do this, if only so that it works for me. If you are amenable I will be happy to send a PR as well. I will test it out on macOS and Windows to make sure it works.

**Comment by Kinneyzhang** [comment]
> Apparently Emacs on macOS uses a terminal shell to run process commands, and Linux and Windows do not.

Thanks for your research. I test it in Majaro, it works well without "-n --heading" in rg, so this may not be the actual reason. But, since you PR can resolve the problem in Windows and Ubuntu, I will merge it. Thanks anyway!

**Comment by aaronbieber** [comment]
Thank you!

I think I figured out the real underlying issue. Because Windows isn't a *NIX system, it has no "shell," so Emacs in Windows spawns processes in some other way.

The Ubuntu test that I did was just running `rg` itself (in WSL), so perhaps Emacs uses a terminal shell in all *NIX systems and Windows is just odd because it doesn't have a concept of a system shell.

Either way, it's definitely a problem in Windows and adding the flags solves it and has no apparent side-effects, so thank you again for the merge.
